File: scripts/manage_lexicon.ts
import 'dotenv/config'; // Load .env file variables first
import pg from 'pg';
import { fileURLToPath } from 'url';
import path from 'path';
import Anthropic, { APIConnectionError, RateLimitError } from '@anthropic-ai/sdk';

const { Pool } = pg;

// Replicate __dirname behavior in ESM
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// --- Anthropic Setup ---
if (!process.env.ANTHROPIC_API_KEY) {
    console.error('CRITICAL: ANTHROPIC_API_KEY environment variable is not set.');
    console.error('Please ensure you have it set in your .env file or environment.');
    process.exit(1);
}
const anthropic = new Anthropic(); // API key is automatically picked up from env var

// Configuration for LLM calls
const LLM_MODEL_EMOTION = "claude-3-haiku-20240307";
const LLM_MODEL_TRANSLATION = "claude-3-haiku-20240307";
const LLM_MAX_RETRIES = 3;
const LLM_RETRY_DELAY_MS = 5000; // milliseconds
const ENGLISH_LANGUAGE_CODE = 'english'; // Corrected based on DB data

// --- Database Setup ---
if (!process.env.DATABASE_URL) {
    console.error('CRITICAL: DATABASE_URL environment variable is not set.');
    console.error('Please ensure you have a .env file with DATABASE_URL=<your_connection_string>');
    process.exit(1);
}

const pool = new Pool({
    connectionString: process.env.DATABASE_URL,
});

// --- Helper Functions ---

/**
 * Simple delay function for retries.
 */
function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Finds the emotion_id for a given emotion name.
 */
async function findEmotionId(client: pg.PoolClient, emotionName: string): Promise<number | null> {
    const result = await client.query('SELECT emotion_id FROM lexicon_emotions WHERE emotion_name = $1', [emotionName.toLowerCase()]);
    return result.rows[0]?.emotion_id ?? null;
}

/**
 * Finds the word_id for a given word text and language code.
 */
async function findWordId(client: pg.PoolClient, wordText: string, languageCode: string): Promise<number | null> {
    const result = await client.query(
        'SELECT word_id FROM lexicon_words WHERE word_text = $1 AND language_code = $2',
        [wordText.toLowerCase(), languageCode.toLowerCase()]
    );
    return result.rows[0]?.word_id ?? null;
}

/**
 * Gets all English words from the database.
 */
async function getAllEnglishWords(client: pg.PoolClient): Promise<{ word_id: number, word_text: string }[]> {
    const result = await client.query('SELECT word_id, word_text FROM lexicon_words WHERE language_code = $1', [ENGLISH_LANGUAGE_CODE]);
    return result.rows;
}

/**
 * Gets all emotion names and IDs from the database.
 */
async function getAllEmotions(client: pg.PoolClient): Promise<{ emotion_id: number, emotion_name: string }[]> {
    const result = await client.query('SELECT emotion_id, emotion_name FROM lexicon_emotions');
    return result.rows;
}

/**
 * Gets all language codes and names from the database, excluding English.
 */
async function getAllNonEnglishLanguages(client: pg.PoolClient): Promise<{ language_code: string, language_name: string }[]> {
    const result = await client.query('SELECT language_code, language_name FROM lexicon_languages WHERE language_code != $1', [ENGLISH_LANGUAGE_CODE]);
    return result.rows;
}

// --- LLM Interaction Functions ---

/**
 * Calls the Anthropic API with retries and error handling.
 */
async function callAnthropicLlm(prompt: string, model: string, max_tokens: number): Promise<string | null> {
    let retries = 0;
    while (retries < LLM_MAX_RETRIES) {
        try {
            const msg = await anthropic.messages.create({
                model: model,
                max_tokens: max_tokens,
                messages: [{ role: 'user', content: prompt }],
            });

            if (msg.content && msg.content.length > 0 && msg.content[0].type === 'text') {
                return msg.content[0].text.trim();
            } else {
                console.warn("  [LLM Warning] Received empty or unexpected content structure from API.");
                return null;
            }
        } catch (error: any) {
            if (error instanceof APIConnectionError) {
                console.error(`  [LLM Error] Anthropic API request failed to connect: ${error}`);
            } else if (error instanceof RateLimitError) {
                console.error(`  [LLM Error] Anthropic API rate limit exceeded: ${error}. Retrying in ${LLM_RETRY_DELAY_MS / 1000}s...`);
                await sleep(LLM_RETRY_DELAY_MS * (retries + 1)); // Exponential backoff might be better
            } else if (error instanceof Anthropic.APIError && error.status) {
                console.error(`  [LLM Error] Anthropic API returned an error status: ${error.status} ${error.message}`);
                // Don't retry on persistent status errors like 4xx
                if (error.status >= 400 && error.status < 500) {
                    return null;
                }
            } else {
                console.error(`  [LLM Error] An unexpected error occurred during API call: ${error}`);
            }
        }
        retries++;
        if (retries < LLM_MAX_RETRIES) {
            console.log(`    Retrying (${retries}/${LLM_MAX_RETRIES})...`);
            await sleep(LLM_RETRY_DELAY_MS * retries); // Basic exponential backoff
        } else {
            console.error("  [LLM Error] Max retries reached.");
        }
    }
    return null;
}

/**
 * Gets a 0/1 emotion score from the LLM for an English word and an emotion dimension.
 */
async function getEmotionScoreFromLlm(word: string, emotionDimension: string): Promise<number | null> {
    const prompt = `Does the English word '${word}' strongly relate to the emotion '${emotionDimension}'? Respond with only the single digit 1 for yes or 0 for no.`;
    console.log(`  [LLM Query] Emotion score for '${word}' / '${emotionDimension}'...`);
    const response = await callAnthropicLlm(prompt, LLM_MODEL_EMOTION, 5); // Small max_tokens for 0/1

    if (response !== null) {
        try {
            const score = parseInt(response, 10);
            if (score === 0 || score === 1) {
                return score;
            } else {
                console.warn(`    [LLM Warning] Invalid score digit received: '${response}'`);
                return null;
            }
        } catch (e) {
            console.warn(`    [LLM Warning] Non-integer score received: '${response}'`);
            return null;
        }
    }
    return null; // API call failed or returned invalid data
}

/**
 * Gets a translation from the LLM for an English word into a target language.
 */
async function getTranslationFromLlm(word: string, languageName: string): Promise<string | null> {
    const plainLanguage = languageName.split(' (')[0]; // Extract plain name if format is "Language (code)"
    const prompt = `What is the most common translation of the English word '${word}' into ${plainLanguage}? Respond with only the single translated word or short phrase. If no direct translation exists or is appropriate, respond with 'N/A'.`;
    console.log(`  [LLM Query] Translation for '${word}' to '${languageName}'...`);
    const response = await callAnthropicLlm(prompt, LLM_MODEL_TRANSLATION, 50); // More tokens for translation

    if (response !== null) {
        if (response.toUpperCase() === 'N/A') {
            console.log(`    [LLM Info] No direct translation found for '${word}' to '${languageName}'.`);
            return null; // Return null for N/A to distinguish from empty string error
        }
        // Basic cleanup: remove potential quotes
        return response.replace(/^["']|["']$/g, '');
    }
    return null; // API call failed
}


// --- Management Functions (Refactored) ---

/**
 * Adds a new custom emotion and uses LLM to associate it with existing English words.
 */
export async function addEmotion(emotionName: string): Promise<void> {
    const client = await pool.connect();
    let newEmotionId: number | null = null;
    const lowerEmotionName = emotionName.toLowerCase();

    try {
        await client.query('BEGIN');

        // 1. Add the emotion
        const result = await client.query(
            `INSERT INTO lexicon_emotions (emotion_name, is_base_nrc)
             VALUES ($1, FALSE)
             ON CONFLICT (emotion_name) DO NOTHING
             RETURNING emotion_id`,
            [lowerEmotionName]
        );

        if ((result.rowCount ?? 0) > 0) {
            newEmotionId = result.rows[0].emotion_id;
            console.log(`Successfully added new emotion '${lowerEmotionName}' with ID: ${newEmotionId}.`);
        } else {
            // Emotion already exists, find its ID
            newEmotionId = await findEmotionId(client, lowerEmotionName);
            if (!newEmotionId) {
                 // Should not happen if ON CONFLICT worked or findEmotionId is correct
                 throw new Error(`Failed to add or find existing emotion '${lowerEmotionName}'.`)
            }
            console.log(`Emotion '${lowerEmotionName}' already exists with ID: ${newEmotionId}. Proceeding to associate words.`);
        }

        // 2. Get all English words
        const englishWords = await getAllEnglishWords(client);
        console.log(`Found ${englishWords.length} English words to process for the new emotion.`);

        // 3. Use LLM to associate words with the new emotion
        let associatedCount = 0;
        for (const word of englishWords) {
            console.log(`Processing word: '${word.word_text}' (ID: ${word.word_id})`);
            const score = await getEmotionScoreFromLlm(word.word_text, lowerEmotionName);

            if (score === 1) {
                console.log(`  -> Associating '${word.word_text}' with '${lowerEmotionName}'...`);
                await client.query(
                    `INSERT INTO word_emotion_associations (word_id, emotion_id)
                     VALUES ($1, $2)
                     ON CONFLICT (word_id, emotion_id) DO NOTHING`,
                    [word.word_id, newEmotionId]
                );
                associatedCount++;
            } else if (score === 0) {
                 console.log(`  -> Score 0, not associating.`);
            } else {
                 console.warn(`  -> Warning: Could not get score for '${word.word_text}'. Skipping association.`);
            }
        }
        console.log(`Associated ${associatedCount} English words with the new emotion '${lowerEmotionName}'.`);

        await client.query('COMMIT');
    } catch (err: any) {
        await client.query('ROLLBACK');
        console.error(`Error adding emotion '${emotionName}' and associating words:`, err.message);
        console.error(err.stack); // Log stack for more detail
        throw err;
    } finally {
        client.release();
    }
}

/**
 * Adds a new English word, uses LLM to associate it with existing emotions,
 * and uses LLM to find and add translations for other languages.
 */
export async function addWord(englishWordText: string): Promise<void> {
    const client = await pool.connect();
    let englishWordId: number | null = null;
    const lowerWordText = englishWordText.toLowerCase();

    try {
        await client.query('BEGIN');

        // 1. Add the English word
        const wordResult = await client.query(
            `INSERT INTO lexicon_words (word_text, language_code)
             VALUES ($1, $2)
             ON CONFLICT (word_text, language_code) DO NOTHING
             RETURNING word_id`,
            [lowerWordText, ENGLISH_LANGUAGE_CODE]
        );

        if ((wordResult.rowCount ?? 0) > 0) {
            englishWordId = wordResult.rows[0].word_id;
            console.log(`Successfully added English word '${lowerWordText}' with ID: ${englishWordId}.`);
        } else {
            englishWordId = await findWordId(client, lowerWordText, ENGLISH_LANGUAGE_CODE);
             if (!englishWordId) {
                 throw new Error(`Failed to add or find existing English word '${lowerWordText}'.`)
            }
            console.log(`English word '${lowerWordText}' already exists with ID: ${englishWordId}. Proceeding with associations and translations.`);
        }

        // 2. Associate with existing emotions via LLM
        const allEmotions = await getAllEmotions(client);
        console.log(`
Associating '${lowerWordText}' with ${allEmotions.length} existing emotions...`);
        let emotionAssociationCount = 0;
        for (const emotion of allEmotions) {
            const score = await getEmotionScoreFromLlm(lowerWordText, emotion.emotion_name);
            if (score === 1) {
                 console.log(`  -> Associating with '${emotion.emotion_name}'...`);
                 await client.query(
                     `INSERT INTO word_emotion_associations (word_id, emotion_id)
                      VALUES ($1, $2)
                      ON CONFLICT (word_id, emotion_id) DO NOTHING`,
                     [englishWordId, emotion.emotion_id]
                 );
                 emotionAssociationCount++;
            } else if (score === 0) {
                 console.log(`  -> Score 0 for '${emotion.emotion_name}', not associating.`);
            } else {
                 console.warn(`  -> Warning: Could not get score for '${emotion.emotion_name}'. Skipping association.`);
            }
        }
         console.log(`Associated '${lowerWordText}' with ${emotionAssociationCount} emotions.`);


        // 3. Get translations for other languages via LLM
        const otherLanguages = await getAllNonEnglishLanguages(client);
        console.log(`
Attempting translations for '${lowerWordText}' into ${otherLanguages.length} other languages...`);
        let translationCount = 0;
        for (const lang of otherLanguages) {
            const translation = await getTranslationFromLlm(lowerWordText, lang.language_name);
            if (translation) {
                const lowerTranslation = translation.toLowerCase();
                 console.log(`  -> Adding translation for ${lang.language_code} (${lang.language_name}): '${lowerTranslation}'`);
                 // Add the translated word to the lexicon_words table
                 await client.query(
                     `INSERT INTO lexicon_words (word_text, language_code)
                      VALUES ($1, $2)
                      ON CONFLICT (word_text, language_code) DO NOTHING`,
                     [lowerTranslation, lang.language_code]
                 );
                 translationCount++;
            } else {
                // Logged within getTranslationFromLlm if N/A or error
            }
        }
        console.log(`Added ${translationCount} translations for '${lowerWordText}'.`);


        await client.query('COMMIT');
    } catch (err: any) {
        await client.query('ROLLBACK');
        console.error(`Error adding word '${englishWordText}' and processing:`, err.message);
        console.error(err.stack);
        throw err;
    } finally {
        client.release();
    }
}


/**
 * Associates an existing word (in a specific language) with an existing emotion.
 * (Kept for manual adjustments if needed)
 */
export async function associateWordEmotion(wordText: string, languageCode: string, emotionName: string): Promise<void> {
    const client = await pool.connect();
    try {
        await client.query('BEGIN');
        const lowerWordText = wordText.toLowerCase();
        const lowerLangCode = languageCode.toLowerCase();
        const lowerEmotionName = emotionName.toLowerCase();

        // Find IDs
        const wordId = await findWordId(client, lowerWordText, lowerLangCode);
        const emotionId = await findEmotionId(client, lowerEmotionName);

        if (!wordId) {
            throw new Error(`Word '${lowerWordText}' (lang: ${lowerLangCode}) not found.`);
        }
        if (!emotionId) {
            throw new Error(`Emotion '${lowerEmotionName}' not found.`);
        }

        // Insert association
        const result = await client.query(
           `INSERT INTO word_emotion_associations (word_id, emotion_id)
            VALUES ($1, $2)
            ON CONFLICT (word_id, emotion_id) DO NOTHING`,
           [wordId, emotionId]
        );

        if ((result.rowCount ?? 0) > 0) {
            console.log(`Successfully associated '${lowerWordText}' (lang: ${lowerLangCode}) with emotion '${lowerEmotionName}'.`);
        } else {
            console.log(`Association between '${lowerWordText}' (lang: ${lowerLangCode}) and '${lowerEmotionName}' already exists.`);
        }
        await client.query('COMMIT');
    } catch (err: any) {
        await client.query('ROLLBACK');
        console.error(`Error associating word '${wordText}' (lang: ${languageCode}) with emotion '${emotionName}':`, err.message);
         throw err;
    } finally {
        client.release();
    }
}

/**
 * Disassociates a word (in a specific language) from an emotion.
 * (Kept for manual adjustments if needed)
 */
export async function disassociateWordEmotion(wordText: string, languageCode: string, emotionName: string): Promise<void> {
    const client = await pool.connect();
    try {
        await client.query('BEGIN');
        const lowerWordText = wordText.toLowerCase();
        const lowerLangCode = languageCode.toLowerCase();
        const lowerEmotionName = emotionName.toLowerCase();

        // Find IDs first to ensure they exist
        const wordId = await findWordId(client, lowerWordText, lowerLangCode);
        const emotionId = await findEmotionId(client, lowerEmotionName);

        if (!wordId) {
            console.warn(`Word '${lowerWordText}' (lang: ${lowerLangCode}) not found. Cannot disassociate.`);
             await client.query('ROLLBACK'); // No changes needed
             return;
        }
        if (!emotionId) {
             console.warn(`Emotion '${lowerEmotionName}' not found. Cannot disassociate.`);
             await client.query('ROLLBACK'); // No changes needed
             return;
        }

        // Delete the association
        const result = await client.query(
            'DELETE FROM word_emotion_associations WHERE word_id = $1 AND emotion_id = $2',
            [wordId, emotionId]
        );

        if ((result.rowCount ?? 0) > 0) {
            console.log(`Successfully disassociated '${lowerWordText}' (lang: ${lowerLangCode}) from emotion '${lowerEmotionName}'.`);
        } else {
            console.log(`No existing association found between '${lowerWordText}' (lang: ${lowerLangCode}) and '${lowerEmotionName}'.`);
        }
        await client.query('COMMIT');
    } catch (err: any) {
        await client.query('ROLLBACK');
        console.error(`Error disassociating word '${wordText}' (lang: ${languageCode}) from emotion '${emotionName}':`, err.message);
         throw err;
    } finally {
        client.release();
    }
}

// --- Command Line Argument Parsing & Execution ---

async function run() {
    const args = process.argv.slice(2); // Remove 'node' and script path
    const command = args[0];

    console.log("Lexicon Management Script (with LLM Integration)");
    console.log("-----------------------------------------------");
    console.log("Usage:");
    console.log("  npm run manage-lexicon -- add-emotion <emotionName>        # Adds emotion and associates existing English words via LLM");
    console.log("  npm run manage-lexicon -- add-word <englishWordText>       # Adds English word, associates with emotions via LLM, gets translations via LLM");
    console.log("  --- Manual Adjustments (Optional) ---");
    console.log("  npm run manage-lexicon -- associate <wordText> <languageCode> <emotionName>");
    console.log("  npm run manage-lexicon -- disassociate <wordText> <languageCode> <emotionName>");
    console.log("-----------------------------------------------");

    try {
        switch (command) {
            case 'add-emotion':
                if (args.length !== 2) throw new Error("Usage: add-emotion <emotionName>");
                await addEmotion(args[1]);
                break;
            case 'add-word':
                 // No longer takes languageCode as it primarily adds English word and finds translations
                if (args.length !== 2) throw new Error("Usage: add-word <englishWordText>");
                await addWord(args[1]);
                break;
            case 'associate':
                if (args.length !== 4) throw new Error("Usage: associate <wordText> <languageCode> <emotionName>");
                await associateWordEmotion(args[1], args[2], args[3]);
                break;
            case 'disassociate':
                if (args.length !== 4) throw new Error("Usage: disassociate <wordText> <languageCode> <emotionName>");
                await disassociateWordEmotion(args[1], args[2], args[3]);
                break;
            default:
                if (command) {
                    console.error(`Unknown command: ${command}`);
                } else {
                     console.log("Please provide a command.");
                }
                // If no command or unknown command, show usage (already printed above)
                 process.exitCode = 1;
                break;
        }
         if (process.exitCode !== 1) { // Don't print success if a command failed or wasn't run
             console.log("\nCommand finished successfully.");
         }
    } catch (error: any) {
        console.error("\nOperation failed:", error.message);
        // Error is already logged within functions, exit code indicates failure
        process.exitCode = 1; // Indicate failure without exiting immediately
    }
}

// Execute the run function and then close the pool
run().finally(async () => {
    // Wait briefly for any outstanding console logs before closing pool
    await sleep(100);
    pool.end(() => console.log('\nDatabase pool closed.'));
}); 

File: corpus_utils/manage_lexicon.py
import csv
import os
import argparse
import sys
import time
import anthropic # Import Anthropic library

# --- Anthropic API Client Initialization --- 
# Ensure ANTHROPIC_API_KEY is set as an environment variable
try:
    client = anthropic.Anthropic()
except anthropic.AnthropicError as e:
    print(f"Error initializing Anthropic client: {e}")
    print("Please ensure the ANTHROPIC_API_KEY environment variable is set correctly.")
    sys.exit(1)

# Configuration for LLM calls
LLM_MODEL_EMOTION = "claude-3-haiku-20240307" # Or another suitable model
LLM_MODEL_TRANSLATION = "claude-3-haiku-20240307" # Can be the same or different
LLM_MAX_RETRIES = 3
LLM_RETRY_DELAY = 5 # seconds

# --- LLM Interaction Functions (Using Anthropic) --- 

def call_anthropic_llm(prompt, model, max_tokens=10): # Reduced max_tokens for score
    """Helper function to call the Anthropic API with retries."""
    retries = 0
    while retries < LLM_MAX_RETRIES:
        try:
            message = client.messages.create(
                model=model,
                max_tokens=max_tokens,
                messages=[
                    {
                        "role": "user",
                        "content": prompt
                    }
                ]
            )
            # Access the content of the first text block
            if message.content and isinstance(message.content, list) and len(message.content) > 0:
                 # Check if the first block is a TextBlock
                 if hasattr(message.content[0], 'text'):
                      return message.content[0].text.strip()
                 else:
                      print(f"  [LLM Warning] Unexpected content block type: {type(message.content[0])}")
                      return None # Or handle other block types if expected
            else:
                 print("  [LLM Warning] Received empty or unexpected content structure from API.")
                 return None
            
        except anthropic.APIConnectionError as e:
            print(f"  [LLM Error] Anthropic API request failed to connect: {e}")
        except anthropic.RateLimitError as e:
            print(f"  [LLM Error] Anthropic API rate limit exceeded: {e}. Retrying in {LLM_RETRY_DELAY}s...")
            time.sleep(LLM_RETRY_DELAY)
        except anthropic.APIStatusError as e:
            print(f"  [LLM Error] Anthropic API returned an error status: {e.status_code} {e.response}")
            # Don't retry on persistent status errors like 4xx
            if 400 <= e.status_code < 500:
                return None
        except Exception as e:
            print(f"  [LLM Error] An unexpected error occurred during API call: {e}")
            
        retries += 1
        if retries < LLM_MAX_RETRIES:
            print(f"    Retrying ({retries}/{LLM_MAX_RETRIES})...")
            time.sleep(LLM_RETRY_DELAY * retries) # Exponential backoff
        else:
            print("  [LLM Error] Max retries reached.")
            
    return None # Return None after exhausting retries or encountering non-retryable errors

def get_emotion_score_from_llm(word, emotion_dimension):
    """Gets a 0/1 score from Anthropic API."""
    prompt = f"Does the English word '{word}' strongly relate to the emotion '{emotion_dimension}'? Respond with only the single digit 1 for yes or 0 for no."
    print(f"  [LLM Query] Emotion score for '{word}' / '{emotion_dimension}'...")
    response = call_anthropic_llm(prompt, LLM_MODEL_EMOTION, max_tokens=5)
    
    if response is not None:
        try:
            score = int(response)
            if score in [0, 1]:
                return score
            else:
                print(f"    [LLM Warning] Invalid score digit received: '{response}'")
                return None
        except ValueError:
            print(f"    [LLM Warning] Non-integer score received: '{response}'")
            return None
    return None # API call failed or returned invalid data

def get_translation_from_llm(word, language_name):
    """Gets a translation from Anthropic API."""
    # Extract plain language name if format is "Language (code)"
    plain_language = language_name.split(' (')[0]
    prompt = f"What is the most common translation of the English word '{word}' into {plain_language}? Respond with only the single translated word or short phrase. If no direct translation exists, respond with 'N/A'."
    print(f"  [LLM Query] Translation for '{word}' to '{language_name}'...")
    response = call_anthropic_llm(prompt, LLM_MODEL_TRANSLATION, max_tokens=50) # More tokens for translations
    
    if response is not None and response.upper() != 'N/A':
        # Basic cleanup: remove potential quotes sometimes added by LLMs
        response = response.strip('"\'')
        return response
    elif response is not None and response.upper() == 'N/A':
        print(f"    [LLM Info] No direct translation found for '{word}' to '{language_name}'.")
        return "" # Return empty string for N/A
    return "" # API call failed or returned None

# --- Generic CSV Handling --- 

def load_csv(filepath):
    """Loads data from a CSV file."""
    if not os.path.exists(filepath):
        print(f"Error: File not found at '{filepath}'")
        return None, None # Return None for header and data
    
    data = []
    header = []
    try:
        with open(filepath, 'r', newline='', encoding='utf-8') as infile:
            reader = csv.reader(infile)
            header = next(reader)
            for row in reader:
                # Basic validation: ensure row has same number of columns as header
                if len(row) == len(header):
                    data.append(row)
                else:
                    print(f"Warning: Skipping malformed row in {filepath} (line {reader.line_num}): expected {len(header)} columns, got {len(row)}")
                    print(f"  Row content: {row}")
        print(f"Successfully loaded {len(data)} rows from {filepath}")
        return header, data
    except Exception as e:
        print(f"Error reading CSV file '{filepath}': {e}")
        sys.exit(1)

def save_csv(filepath, header, data):
    """Saves data to a CSV file atomically."""
    temp_filepath = filepath + ".tmp"
    try:
        with open(temp_filepath, 'w', newline='', encoding='utf-8') as outfile:
            writer = csv.writer(outfile)
            writer.writerow(header)
            writer.writerows(data)
        
        os.replace(temp_filepath, filepath)
        print(f"Successfully saved updated data to {filepath}")
        return True
    except Exception as e:
        print(f"Error writing CSV file '{filepath}': {e}")
        if os.path.exists(temp_filepath):
            try: os.remove(temp_filepath) 
            except OSError: pass
        return False

# --- Core Logic (Updated for Two Files) --- 

def add_emotion_dimension(emotions_filepath, new_emotion):
    """Adds a new emotion dimension column to the emotions CSV."""
    header, data = load_csv(emotions_filepath)
    if header is None:
        sys.exit(1)

    if new_emotion in header:
        print(f"Error: Emotion dimension '{new_emotion}' already exists in {emotions_filepath}.")
        sys.exit(1)

    print(f"Adding new emotion dimension '{new_emotion}' to {emotions_filepath}")
    header.append(new_emotion)
    new_emotion_col_index = len(header) - 1

    try:
        word_index = header.index('English Word')
    except ValueError:
        print(f"Error: 'English Word' column not found in {emotions_filepath}.")
        sys.exit(1)

    updated_data = []
    total_rows = len(data)
    for i, row in enumerate(data):
        word = row[word_index]
        print(f"Processing word {i+1}/{total_rows}: '{word}'")
        
        score = get_emotion_score_from_llm(word, new_emotion)
        
        if score is None:
             print(f"  -> Warning: Could not get score for '{word}'. Setting to 0.")
             score_str = "0" 
        else:
             print(f"  -> Score: {score}")
             score_str = str(score)

        # Ensure row is long enough before appending (though load_csv should handle this)
        if len(row) < new_emotion_col_index:
             row.extend(['0'] * (new_emotion_col_index - len(row))) # Pad with default '0'
             
        row.append(score_str)
        updated_data.append(row)

    save_csv(emotions_filepath, header, updated_data)

def add_word(emotions_filepath, translations_filepath, new_word):
    """Adds a new word to both emotions and translations CSVs."""
    emo_header, emo_data = load_csv(emotions_filepath)
    trans_header, trans_data = load_csv(translations_filepath)

    if emo_header is None or trans_header is None:
        sys.exit(1) # Error already printed by load_csv

    # Find word column index (should be 0 in both)
    try:
        emo_word_index = emo_header.index('English Word')
        trans_word_index = trans_header.index('English Word')
        if emo_word_index != 0 or trans_word_index != 0:
             print("Warning: 'English Word' column is not the first column in one or both files. Proceeding, but check file structure.")
    except ValueError:
        print("Error: 'English Word' column not found in one or both CSV files.")
        sys.exit(1)

    # Check if word already exists (case-insensitive check in emotions file)
    existing_words = {row[emo_word_index].lower() for row in emo_data}
    if new_word.lower() in existing_words:
        print(f"Error: Word '{new_word}' already exists in {emotions_filepath}.")
        sys.exit(1)

    print(f"Adding new word: '{new_word}' to both files.")

    # --- Process Emotions File --- 
    print("\nProcessing Emotions...")
    new_emo_row = [''] * len(emo_header)
    new_emo_row[emo_word_index] = new_word
    emotion_columns = [(col_name, i) for i, col_name in enumerate(emo_header) if i != emo_word_index]
    print(f"Getting scores for {len(emotion_columns)} emotion dimensions...")
    
    for emotion, index in emotion_columns:
        score = get_emotion_score_from_llm(new_word, emotion)
        if score is None:
             print(f"  -> Warning: Could not get score for '{emotion}'. Setting to 0.")
             score_str = "0"
        else:
             print(f"  -> Score for '{emotion}': {score}")
             score_str = str(score)
        new_emo_row[index] = score_str
        
    emo_data.append(new_emo_row)

    # --- Process Translations File --- 
    print("\nProcessing Translations...")
    new_trans_row = [''] * len(trans_header)
    new_trans_row[trans_word_index] = new_word
    language_columns = [(col_name, i) for i, col_name in enumerate(trans_header) if i != trans_word_index]
    print(f"Getting translations for {len(language_columns)} languages...")
    
    for language, index in language_columns:
        translation = get_translation_from_llm(new_word, language)
        if not translation:
             print(f"  -> Info: No translation obtained for '{language}'. Leaving empty.")
             translation_str = ""
        else:
             # Limit length just in case LLM is verbose
             translation_str = translation[:100] # Limit translation length
             print(f"  -> Translation for '{language}': {translation_str}")
             
        new_trans_row[index] = translation_str

    trans_data.append(new_trans_row)

    # --- Save Both Files --- 
    print("\nSaving updated files...")
    emo_saved = save_csv(emotions_filepath, emo_header, emo_data)
    trans_saved = save_csv(translations_filepath, trans_header, trans_data)

    if not emo_saved or not trans_saved:
        print("Error: Failed to save one or both files. Check previous error messages.")
        # Note: This could leave files in an inconsistent state if one saves and the other fails.
        # More robust implementation might involve backups or transactional saving.
        sys.exit(1)

# --- Main Execution --- 

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Manage the NRC Emotion and Translation Lexicon CSVs using Anthropic API.")
    parser.add_argument("--emotions-file", required=True, help="Path to the nrc_lexicon_emotions.csv file.")
    parser.add_argument("--translations-file", required=True, help="Path to the nrc_lexicon_translations.csv file.")
    
    mode_group = parser.add_mutually_exclusive_group(required=True)
    mode_group.add_argument("--add-emotion", metavar="EMOTION_NAME", help="Add a new emotion dimension to the emotions file.")
    mode_group.add_argument("--add-word", metavar="NEW_WORD", help="Add a new English word to both files.")

    args = parser.parse_args()

    # Basic file existence check (load_csv handles more specific errors)
    if not os.path.exists(args.emotions_file):
        print(f"Error: Emotions file not found at '{args.emotions_file}'")
        sys.exit(1)
    if not os.path.exists(args.translations_file):
        print(f"Error: Translations file not found at '{args.translations_file}'")
        sys.exit(1)
        
    if args.add_emotion:
        add_emotion_dimension(args.emotions_file, args.add_emotion)
    elif args.add_word:
        add_word(args.emotions_file, args.translations_file, args.add_word)

    print("\nScript finished.") 